actix-http/src/client/connection.rs

```diff
@@ -19,177 +21,230 @@ use super::error::SendRequestError;
+        if close {
+            self.close();
+        } else {
+            self.release();
+        }
+    }
 
-    /// Send request and body
-    fn send_request<B, H>(
-        self,
-        head: H,
-        body: B,
-    ) -> LocalBoxFuture<'static, Result<(ResponseHead, Payload), SendRequestError>>
-    where
-        B: MessageBody + 'static,
-        H: Into<RequestHeadType> + 'static;
+    /// Close connection
+    fn close(&mut self) {
+        let io = self.io.take().unwrap();
+        self.acquired.close(ConnectionInnerType::H2(io));
+    }
 
-    /// Send request, returns Response and Framed
-    fn open_tunnel<H: Into<RequestHeadType> + 'static>(
-        self,
-        head: H,
-    ) -> LocalBoxFuture<
-        'static,
-        Result<(ResponseHead, Framed<Self::Io, ClientCodec>), SendRequestError>,
-    >;
+    /// Release this connection to the connection pool
+    fn release(&mut self) {
+        let io = self.io.take().unwrap();
+        self.acquired
+            .release(ConnectionInnerType::H2(io), self.created);
+    }
 }
 
-#[doc(hidden)]
-/// HTTP client connection
-pub struct IoConnection<T>
-where
-    T: AsyncWrite + Unpin + 'static,
-{
-    io: Option<ConnectionType<T>>,
-    created: time::Instant,
-    pool: Acquired<T>,
+/// `H2ConnectionInner` has two parts: `SendRequest` and `Connection`.
+///
+/// `Connection` is spawned as an async task on runtime and `H2ConnectionInner` holds a handle
+/// for this task. Therefore, it can wake up and quit the task when SendRequest is dropped.
+pub(super) struct H2ConnectionInner {
+    handle: JoinHandle<()>,
+    sender: SendRequest<Bytes>,
 }
 
-impl<T> fmt::Debug for IoConnection<T>
+impl H2ConnectionInner {
+    pub(super) fn new<Io: ConnectionIo>(
+        sender: SendRequest<Bytes>,
+        connection: h2::client::Connection<Io>,
+    ) -> Self {
+        let handle = actix_rt::spawn(async move {
+            let _ = connection.await;
+        });
+
+        Self { handle, sender }
+    }
+}
+
+/// Cancel spawned connection task on drop.
+impl Drop for H2ConnectionInner {
+    fn drop(&mut self) {
+        self.handle.abort();
+    }
+}
+
+#[allow(dead_code)]
+/// Unified connection type cover Http1 Plain/Tls and Http2 protocols
+pub enum Connection<A, B = Box<dyn ConnectionIo>>
 where
-    T: AsyncWrite + Unpin + fmt::Debug + 'static,
+    A: ConnectionIo,
+    B: ConnectionIo,
 {
-    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        match self.io {
-            Some(ConnectionType::H1(ref io)) => write!(f, "H1Connection({:?})", io),
-            Some(ConnectionType::H2(_)) => write!(f, "H2Connection"),
-            None => write!(f, "Connection(Empty)"),
-        }
-    }
+    Tcp(ConnectionType<A>),
+    Tls(ConnectionType<B>),
 }
 
-impl<T: AsyncRead + AsyncWrite + Unpin> IoConnection<T> {
-    pub(crate) fn new(
-        io: ConnectionType<T>,
+/// Unified connection type cover Http1/2 protocols
+pub enum ConnectionType<Io: ConnectionIo> {
```

Comment:
There is no public mod in client module and this type is not exported. It's marked as public because some public service impl return it as output.

The only connection type export is `Connection` which can only be the two variant it has(I guess)

Reply:
as an output, if you could match on the result then it should be non_exhaustive

Reply:
How can it be matched if the type itself is not importable?

Reply:
truth